### PR TITLE
Extract dependencies from go mod graph if is available

### PR DESCRIPTION
**Pull request by tonyredondo**
Extract dependencies from `go mod graph` if is available.

### CODE REVIEW

File: agent/dependencies.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"os/exec"
+	"strings"
+
+	"go.undefinedlabs.com/scopeagent/tags"
+)
+
+// Gets the dependencies map
+func getDependenciesMap() map[string]interface{} {
```
Comment (fermayo) [inline agent/dependencies.go]:
```suggestion
func getDependencyMap() map[string]string {
```

File: agent/dependencies.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	dependencies := map[string]interface{}{}
```
Comment (fermayo) [inline agent/dependencies.go]:
```suggestion
	dependencies := map[string]string{}
```

File: agent/dependencies.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+		deps := map[string]string{}
+		for _, v := range lines {
+			if len(v) == 0 {
+				continue
+			}
+			lIdx := strings.LastIndex(v, " ") + 1
+			arr := strings.Split(v[lIdx:], "@")
+			if len(arr) == 2 {
+				deps[arr[0]] = arr[1]
+			}
```
Comment (fermayo) [inline agent/dependencies.go]: This could be rewritten with a simple regex

File: agent/agent.go [hunk 4]
```diff
@@ -196,6 +196,9 @@ func NewAgent(options ...Option) (*Agent, error) {
 
 	agent.metadata[tags.InContainer] = isRunningInContainer()
 
+	// Dependencies
+	addToMapIfEmpty(agent.metadata, getDependenciesMap())
```
Comment (fermayo) [inline agent/agent.go]:
```suggestion
	agent.metadata[tags.Dependencies] = getDependencyMap()
```

File: agent/dependencies.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	"go.undefinedlabs.com/scopeagent/tags"
+)
+
+// Gets the dependencies map
+func getDependenciesMap() map[string]interface{} {
+	dependencies := map[string]interface{}{}
+	if modGraphBytes, err := exec.Command("go", "mod", "graph").Output(); err == nil {
+		strGraph := string(modGraphBytes)
+		lines := strings.Split(strGraph, "\n")
+		deps := map[string]string{}
```
Comment (fermayo) [inline agent/dependencies.go]:
```suggestion
```

File: agent/dependencies.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+		for _, v := range lines {
+			if len(v) == 0 {
+				continue
+			}
+			lIdx := strings.LastIndex(v, " ") + 1
+			arr := strings.Split(v[lIdx:], "@")
+			if len(arr) == 2 {
+				deps[arr[0]] = arr[1]
```
Comment (fermayo) [inline agent/dependencies.go]:
```suggestion
				dependencies[arr[0]] = arr[1]
```

File: agent/dependencies.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+			}
+		}
+		dependencies[tags.Dependencies] = deps
```
Comment (fermayo) [inline agent/dependencies.go]:
```suggestion
```

File: agent/agent_test.go [hunk 8]
```diff
@@ -32,3 +34,11 @@ func TestDsnParser(t *testing.T) {
...
 	}
 }
+
+func TestGetDependencies(t *testing.T) {
+	deps := getDependenciesMap()
+	fmt.Printf("%v\n", deps)
+	if _, ok := deps[tags.Dependencies]; !ok {
+		t.Fail()
+	}
+}
```
Comment (fermayo) [inline agent/agent_test.go]: I'd rewrite this test to align with my proposed function signature

File: agent/dependencies.go (hunk 3 above)
Comment (tonyredondo) [reply agent/dependencies.go]: Done.

**Comment by tonyredondo** [comment]
All changes were applied

File: agent/agent_test.go [hunk 9]
```diff
@@ -32,3 +32,12 @@ func TestDsnParser(t *testing.T) {
 		})
 	}
 }
+
+func TestGetDependencies(t *testing.T) {
+	deps := getDependencyMap()
+	fmt.Printf("%v\n", deps)
+	fmt.Printf("%d\n", len(deps))
+	if len(deps) < 40 {
```
Comment (fermayo) [inline agent/agent_test.go]: Instead of a random number here, could we read the number of lines in `go.mod`?

Comment (tonyredondo) [reply agent/agent_test.go]: `go.mod` contains only the direct dependencies of the agent, when you run `go mod graph` resolves all dependencies, and outputs indirect dependencies too, In this example go.mod has 16 dependencies inside the `require` and `go mod graph` returns 43 dependencies (direct+indirect).
At the end all dependencies are available for the compiler + linker. If this map is going to be used in the RCA I think more is better.

For this test, i can run `go mod graph` and count the number of lines.

**Comment by tonyredondo** [comment]
@fermayo In the last commit I added support for multiple versions of the same dependency due indirect dependencies, and did a better test for the regex

**Pull request merged**